Automattic/simplenote-ios
# Implements ResultsController

### Details:
In this PR we're encapsulating (nearly all) of the NSFetchedResultsController interactions within a single class, to be named **ResultsController**.

This is one of many PR(s). Since the AdvancedSearch feature requires to mix results from multiple Entities, we're likely to use a **DataSource**, which will interact with **ResultsController**.

cc @bummytime (Thanks in advance Bummy!!)
Ref. #507

### Notes:
Live refresh of remote changes coming up in #573!

### Test: List Filters
1. Fresh install the app
2. Log into your account

- [ ] Verify that while your notes are downloading, there's a spinner on top
- [ ] Verify **All Notes** effectively display all of your notes
- [ ] Verify **Trash** renders the trashed notes
- [ ] Verify **Untagged Notes** filters out just the notes that don't have a valid tag
- [ ] Verify the App's title is updated in each one of the above scenarios
- [ ] Verify that the Trash shows a button that reads `Empty` on the top right corner
- [ ] Verify the Sort Mode setting is also always respected

### Test: List Actions
1. Log into your account

- [ ] Verify that 3d Touch is available for every note (except Trashed Notes)
- [ ] Verify that Swipeable actions work correctly (for Deleted and non Deleted notes)
- [ ] Verify that pressing over any Note cell pushes the Editor

### Test: Searching
1. Log into your account
2. Press the Search Bar
3. Type any random keyword

- [ ] Verify that searching by any given keyword filters out the related notes
- [ ] Verify that any matching keywords are highlighted
- [ ] Verify that pressing over any of the results row pushes the Editor in search mode (a bar should show up at the bottom)

### Release:
These changes do not require release notes.
